server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs

```diff
@@ -49,54 +53,77 @@ import           Hasura.Prelude
 import           Hasura.RQL.Types
 import           Hasura.SQL.Value
 
-newtype RemoteRelKey =
-  RemoteRelKey Int
-  deriving (Eq, Ord, Show, Hashable)
-
+-- | Like a 'VQ.Field' but with a path indicating where its response should be
+-- spliced back into the hasura result.
 data RemoteRelField =
   RemoteRelField
     { rrRemoteField   :: !RemoteField
+    -- ^ The hasura to remote schema field mapping.
     , rrArguments     :: !ArgsMap
+    -- ^ See '_fArguments'.
     , rrSelSet        :: !SelSet
+    -- ^ Selection set of 'rrAlias' (may be empty)
     , rrRelFieldPath  :: !RelFieldPath
+    -- ^ Splice path.
     , rrAlias         :: !G.Alias
+    -- ^ A possible alias (provided by client) for `rtrName .
+    -- rmfRemoteRelationship . rrRemoteField`, otherwise this will be defaulted
+    -- to that value.
     , rrAliasIndex    :: !Int
+    -- ^ index of selection set at 'rrRelFieldPath' at which this remote field should be spliced.
+    -- This is morally the last component of 'rrRelFieldPath'.
     , rrPhantomFields :: ![Text]
+    -- ^ Hasura fields which are not in the selection set, but are required as
+    -- parameters to satisfy the remote join.
     }
   deriving (Show)
 
+-- | A path of ordered keys (suitable for 'OJ.insert') allowing us to splice a
+-- remote result at a nested location.
 newtype RelFieldPath =
   RelFieldPath {
     unRelFieldPath :: (Seq.Seq (Int, G.Alias))
   } deriving (Show, Monoid, Semigroup, Eq)
 
-data InsertPath =
-  InsertPath
-    { ipFields :: !(Seq.Seq Text)
-    , ipIndex  :: !(Maybe ArrayIndex)
-    }
-  deriving (Show, Eq)
-
-newtype ArrayIndex =
-  ArrayIndex Int
-  deriving (Show, Eq, Ord)
-
--- Rebuild the field with remote relationships removed, and paths that
--- point back to them.
-rebuildFieldStrippingRemoteRels ::
-     VQ.Field -> Maybe (VQ.Field, NonEmpty RemoteRelField)
+-- | Rebuild the query tree, returning a new tree with remote relationships
+-- removed, along with a list of paths that point back to them.
+--
+-- 'Nothing' means there were no remote fields.
+--
+-- Output here is eventually assembled into a 'QExecPlan' 'Tree' in
+-- 'getExecPlan', and finally executed in 'runGQ'.
+rebuildFieldStrippingRemoteRels
+  :: VQ.Field 
+  -> Maybe (VQ.Field, NonEmpty RemoteRelField)
+  -- ^ NOTE: the ordering of the _fSelSet fields in the result 'VQ.Field' seems not to matter at 
+  -- all for correctness here (based on experimentation)... I haven't puzzled out why yet.
```

Comment:
How are you testing the ordering of fields? The python test suite does not account for ordering so this might have failed to show there.

Reply:
Right, I did a temporary rebase on my branch here where I can assert ordering: https://github.com/hasura/graphql-engine/pull/2831  (this is one of the PRs I worked on but set aside).